File: VB/Helper/FilterCriteriaHelper.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Collections.Generic
Imports System.ComponentModel
Imports System.Data
Imports System.Drawing
Imports System.Text
Imports System.Windows.Forms
Imports DevExpress.XtraGrid.Views.Grid
Imports DevExpress.Data.Filtering
Imports DevExpress.Data.Helpers
Imports DevExpress.XtraGrid.Columns
Imports DevExpress.Data
Imports DevExpress.XtraEditors.Repository
Imports DevExpress.XtraGrid.Views.Base
Imports DevExpress.Data.Filtering.Helpers
Imports System.Collections

Namespace WindowsApplication1
	Public NotInheritable Class FilterCriteriaHelper

		Private Sub New()
		End Sub
		Public Shared Function ReplaceFilterCriteria(ByVal source As CriteriaOperator, ByVal prevOperand As CriteriaOperator, ByVal newOperand As CriteriaOperator) As CriteriaOperator
			Dim groupOperand As GroupOperator = TryCast(source, GroupOperator)
			If ReferenceEquals(groupOperand, Nothing) Then
				Return newOperand
			End If
			Dim clone As GroupOperator = groupOperand.Clone()
			clone.Operands.Remove(prevOperand)
			If clone.Equals(source) Then
				Return newOperand
			End If
			clone.Operands.Add(newOperand)
			Return clone
		End Function

		Public Shared Function ReplaceFindPanelCriteria(ByVal source As CriteriaOperator, ByVal view As GridView, ByVal newOperand As CriteriaOperator) As CriteriaOperator
			Return ReplaceFilterCriteria(source, ConvertFindPanelTextToCriteriaOperator(view.FindFilterText, view, True), newOperand)
		End Function

		Public Shared Function ConvertFindPanelTextToCriteriaOperator(ByVal findPanelText As String, ByVal view As GridView, ByVal applyPrefixes As Boolean) As CriteriaOperator
			If (Not String.IsNullOrEmpty(findPanelText)) Then
				Dim parseResult As FindSearchParserResults = New FindSearchParser().Parse(findPanelText, GetFindToColumnsCollection(view))
				If applyPrefixes Then
					parseResult.AppendColumnFieldPrefixes()
				End If

				Return DxFtsContainsHelperAlt.Create(parseResult, FilterCondition.Contains, False)
			End If
			Return Nothing
		End Function



		Private Shared Function GetFindToColumnsCollection(ByVal view As GridView) As ICollection
			Dim mi As System.Reflection.MethodInfo = GetType(ColumnView).GetMethod("GetFindToColumnsCollection", System.Reflection.BindingFlags.NonPublic Or System.Reflection.BindingFlags.Instance)
			Return TryCast(mi.Invoke(view, Nothing), ICollection)
		End Function

		Public Shared Function MyConvertFindPanelTextToCriteriaOperator(ByVal view As GridView) As CriteriaOperator
			Return ConvertFindPanelTextToCriteriaOperator(String.Format("""{0}""", view.FindFilterText), view, False)
		End Function
	End Class
End Namespace

File: CS/E2801/WindowsApplication1/Helper/FilterCriteriaHelper.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Text;
using System.Windows.Forms;
using DevExpress.XtraGrid.Views.Grid;
using DevExpress.Data.Filtering;
using DevExpress.Data.Helpers;
using DevExpress.XtraGrid.Columns;
using DevExpress.Data;
using DevExpress.XtraEditors.Repository;
using DevExpress.XtraGrid.Views.Base;
using DevExpress.Data.Filtering.Helpers;
using System.Collections;

namespace WindowsApplication1
{
    public static class FilterCriteriaHelper
    {

        public static CriteriaOperator ReplaceFilterCriteria(CriteriaOperator source, CriteriaOperator prevOperand, CriteriaOperator newOperand)
        {
            GroupOperator groupOperand = source as GroupOperator;
            if (ReferenceEquals(groupOperand, null))
                return newOperand;
            GroupOperator clone = groupOperand.Clone();
            clone.Operands.Remove(prevOperand);
            if (clone.Equals(source))
                return newOperand;
            clone.Operands.Add(newOperand);
            return clone;
        }

        public static CriteriaOperator ReplaceFindPanelCriteria(CriteriaOperator source, GridView view, CriteriaOperator newOperand)
        {
            return ReplaceFilterCriteria(source, ConvertFindPanelTextToCriteriaOperator(view.FindFilterText, view, true), newOperand);
        }

        public static CriteriaOperator ConvertFindPanelTextToCriteriaOperator(string findPanelText, GridView view, bool applyPrefixes)
        {
            if (!string.IsNullOrEmpty(findPanelText))
            {
                FindSearchParserResults parseResult = new FindSearchParser().Parse(findPanelText, GetFindToColumnsCollection(view));
                if (applyPrefixes)
                    parseResult.AppendColumnFieldPrefixes();
                return DxFtsContainsHelper.Create(parseResult);
            }
            return null;
        }


        private static ICollection GetFindToColumnsCollection(GridView view)
        {
            System.Reflection.MethodInfo mi = typeof(ColumnView).GetMethod("GetFindToColumnsCollection", System.Reflection.BindingFlags.NonPublic | System.Reflection.BindingFlags.Instance);
            return mi.Invoke(view, null) as ICollection;
        }

        public static CriteriaOperator MyConvertFindPanelTextToCriteriaOperator(GridView view)
        {
            return ConvertFindPanelTextToCriteriaOperator(String.Format("\"{0}\"", view.FindFilterText), view, false);
        }
    }
}